### ISSUE TITLE
[BUG] Can't parse PDF, no thumbnail created. Missing Arial font?

### ISSUE THREAD

**Issue by Fedalto** [paperless-ngx/paperless-ngx#277]
I'm reopening https://github.com/jonaswinkler/paperless-ng/issues/1140, opened by @jhf2442, here.
I also have this issue with 1 PDF file for now, with the same error logs and behavior that @jhf2442 reported.
I upgraded to paperless-ngx and could verify that this bug still happens.

Below is a copy/paste of the original https://github.com/jonaswinkler/paperless-ng/issues/1140 issue:

**Describe the bug**
Out of 1600 PDFs consumed so far, 4 PDFs fail to be properly displayed after ingestion

- PNG is 267 bytes in size, empty (white background) PNG, 500*704px (checked with gimp). Running the convert/optipng commands manually on my arch linux system leads to perfect thumbnail
-    Preview as well as "edit" window display blank white page instead of actual PDF document
-    Content tab displays OCR'ed text -> OCR is fine
-    Download of the original PDF returns OK PDF file

3 documents are from the same amazon seller, but the 4th not. PDF properties also show a different creator program

**To Reproduce**

 1.   Go to '...'
1.    Click on '....'
1.    Scroll down to '....'
1.    See error

**Expected behavior**

**Screenshots**

**Webserver logs**

Relevant part (IMHO) is the CID font "Arial" stuff -> how to install this in the docker container ?
```
webserver_1  | [2021-06-26 22:16:14,473] [INFO] [paperless.consumer] Consuming 2018-03-13_Rechnung.pdf
webserver_1  | [2021-06-26 22:16:15,186] [ERROR] [ocrmypdf._exec.ghostscript] GPL Ghostscript 9.27 (2019-04-04)
webserver_1  | Copyright (C) 2018 Artifex Software, Inc.  All rights reserved.                           
webserver_1  | This software is supplied under the GNU AGPLv3 and comes with NO WARRANTY:                
webserver_1  | see the file COPYING for details.                                                         
webserver_1  | GPL Ghostscript 9.27: UTF16BE text string detected in DOCINFO cannot be represented in XMP for PDF/A1, discarding DOCINFO
webserver_1  | GPL Ghostscript 9.27: UTF16BE text string detected in DOCINFO cannot be represented in XMP for PDF/A1, discarding DOCINFO                             
webserver_1  | GPL Ghostscript 9.27: UTF16BE text string detected in DOCINFO cannot be represented in XMP for PDF/A1, discarding DOCINFO                       
webserver_1  | GPL Ghostscript 9.27: UTF16BE text string detected in DOCINFO cannot be represented in XMP for PDF/A1, discarding DOCINFO
webserver_1  | GPL Ghostscript 9.27: UTF16BE text string detected in DOCINFO cannot be represented in XMP for PDF/A1, discarding DOCINFO
webserver_1  | Processing pages 1 through 1.                                                             
webserver_1  | Page 1                                                                                    
webserver_1  | Can't find CID font "Arial".                                                              
webserver_1  | Attempting to substitute CID font /Adobe-Identity for /Arial, see doc/Use.htm#CIDFontSubstitution.
webserver_1  | The substitute CID font "Adobe-Identity" is not provided either. attempting to use fallback CIDFont.See doc/Use.htm#CIDFontSubstitution.
webserver_1  | The fallback CID font "CIDFallBack" is not provided.  Finally attempting to use ArtifexBullet.
webserver_1  |                                                                                           
webserver_1  | [2021-06-26 22:16:15,186] [ERROR] [ocrmypdf._exec.ghostscript]  Error reading a content stream. The page may be incomplete.
webserver_1  |                Output may be incorrect.                                                   
webserver_1  |                                                                                           
webserver_1  | [2021-06-26 22:16:15,186] [ERROR] [ocrmypdf._exec.ghostscript]  Error: File did not complete the page properly and may be damaged.
webserver_1  |                Output may be incorrect.                                                   
webserver_1  | GPL Ghostscript 9.27: Setting Overprint Mode to 1                                         
webserver_1  |  not permitted in PDF/A-2, overprint mode not set                                   
webserver_1  |                                                                                           
webserver_1  |                                                                                           
webserver_1  | [2021-06-26 22:16:15,438] [WARNING] [paperless.parsing.tesseract] Encountered an error while running OCR: No text was found in the original document. Attempting force OCR to get the text.
webserver_1  | [2021-06-26 22:16:15,783] [ERROR] [ocrmypdf._exec.ghostscript]    **** Error reading a content stream. The page may be incomplete.
webserver_1  |                Output may be incorrect.                                                   
webserver_1  |    **** Error: File did not complete the page properly and may be damaged.                
webserver_1  |                Output may be incorrect.                                             
```

**Relevant information**

-    Host OS of the machine running paperless: [Ubuntu 20.04]
-    Browser [chrome]
-    Version [e.g. 1.4.5]
-    Installation method: [docker]
-    Any configuration changes you made in `docker-compose.yml`, `docker-compose.env` or `paperless.conf`.

**Comment by jhf2442** [comment]
Just switched to latest/greatest ngx container - problem remains.
I've sent the PDF privately to yschroeder, maybe he can find something out

**Comment by jhf2442** [comment]
btw, if somebody could give me a kickstart, I could try investigating by myself -> in which file to start looking for the ghostscript stuff ?

**Comment by sturaipo** [comment]
I can reproduce the error as well.
In this case these are the payslips from my employer 
here is the error log:
```

[2022-03-22 19:49:33,328] [INFO] [paperless.consumer] Consuming IEFPY_324078_28_05_21.pdf


[2022-03-22 19:49:33,878] [ERROR] [ocrmypdf._exec.ghostscript] GPL Ghostscript 9.53.3 (2020-10-01)


Copyright (C) 2020 Artifex Software, Inc.  All rights reserved.


This software is supplied under the GNU AGPLv3 and comes with NO WARRANTY:


see the file COPYING for details.


Processing pages 1 through 1.


Page 1


Can't find CID font "Arial".


Attempting to substitute CID font /Adobe-Identity for /Arial, see doc/Use.htm#CIDFontSubstitution.


The substitute CID font "Adobe-Identity" is not provided either. attempting to use fallback CIDFont.See doc/Use.htm#CIDFontSubstitution.


The fallback CID font "CIDFallBack" is not provided.  Finally attempting to use ArtifexBullet.


   


[2022-03-22 19:49:33,879] [ERROR] [ocrmypdf._exec.ghostscript]  Error reading a content stream. The page may be incomplete.


               Output may be incorrect.


   


[2022-03-22 19:49:33,879] [ERROR] [ocrmypdf._exec.ghostscript]  Error: File did not complete the page properly and may be damaged.


               Output may be incorrect.



[2022-03-22 19:49:33,945] [WARNING] [paperless.parsing.tesseract] Encountered an error while running OCR: No text was found in the original document. Attempting force OCR to get the text.


[2022-03-22 19:49:34,238] [ERROR] [ocrmypdf._exec.ghostscript]    **** Error reading a content stream. The page may be incomplete.


               Output may be incorrect.


   **** Error: File did not complete the page properly and may be damaged.


               Output may be incorrect.



[2022-03-22 19:49:35,417] [ERROR] [ocrmypdf._exec.tesseract] [tesseract] Error during processing.


[2022-03-22 19:49:35,984] [ERROR] [ocrmypdf._exec.ghostscript]    **** Error reading a content stream. The page may be incomplete.


               Output may be incorrect.


   **** Error: File did not complete the page properly and may be damaged.


               Output may be incorrect.



[2022-03-22 19:49:51,008] [INFO] [paperless.handlers] Assigning document type bank statement to 2022-03-22 IEFPY_324078_28_05_21


[2022-03-22 19:49:51,677] [INFO] [paperless.consumer] Document 2022-03-22 IEFPY_324078_28_05_21 consumption finished
```

I can also help debugging this

**Comment by b1n4ryj4n** [comment]
i have the same issue but for the font ArialMT

`[2022-03-25 19:08:46,935] [ERROR] [ocrmypdf._exec.ghostscript] GPL Ghostscript 9.50 (2019-10-15)
Copyright (C) 2019 Artifex Software, Inc.  All rights reserved.
This software is supplied under the GNU AGPLv3 and comes with NO WARRANTY:
see the file COPYING for details.
Processing pages 1 through 1.
Page 1
Can't find CID font "ArialMT".
Attempting to substitute CID font /Adobe-Identity for /ArialMT, see doc/Use.htm#CIDFontSubstitution.
The substitute CID font "Adobe-Identity" is not provided either. attempting to use fallback CIDFont.See doc/Use.htm#CIDFontSubstitution.
The fallback CID font "CIDFallBack" is not provided.  Finally attempting to use ArtifexBullet.

[2022-03-25 19:08:46,935] [ERROR] [ocrmypdf._exec.ghostscript]  Error reading a content stream. The page may be incomplete.
               Output may be incorrect.

[2022-03-25 19:08:46,935] [ERROR] [ocrmypdf._exec.ghostscript]  Error: File did not complete the page properly and may be damaged.
               Output may be incorrect.
GPL Ghostscript 9.50: Setting Overprint Mode to 1
 not permitted in PDF/A-2, overprint mode not set
[2022-03-25 19:08:47,015] [WARNING] [paperless.parsing.tesseract] Encountered an error while running OCR: No text was found in the original document. Attempting force OCR to get the text.
[2022-03-25 19:08:47,382] [ERROR] [ocrmypdf._exec.ghostscript]    **** Error reading a content stream. The page may be incomplete.
               Output may be incorrect.
   **** Error: File did not complete the page properly and may be damaged.
               Output may be incorrect.
[2022-03-25 19:08:48,772] [ERROR] [ocrmypdf._exec.tesseract] [tesseract] Error during processing.
[2022-03-25 19:08:49,528] [ERROR] [ocrmypdf._exec.ghostscript]    **** Error reading a content stream. The page may be incomplete.
               Output may be incorrect.
   **** Error: File did not complete the page properly and may be damaged.
               Output may be incorrect.
[2022-03-25 19:09:03,437] [INFO] [paperless.consumer] Document 2018-03-12 SCAN_20180503_0001 consumption finished
19:09:03 [Q] INFO Process-1:40 stopped doing work
19:09:03 [Q] INFO Processed [SCAN_20180503_0001.pdf]`

**Comment by stumpylog** [comment]
This does appear to be a missing font used in the particular pdf, based on stumbling into this area of ghostscript: https://ghostscript.com/doc/9.50/Use.htm#CIDFonts

Not sure yet if there is just a package which can be installed to fix it

**Comment by sturaipo** [comment]
Some updates about the pdf file I am fighting with.

The PDF file uses fonts Arial and Arial,Bold which are not embedded (as they should be, but I have no control about it).

Here is the output of pdffont
```
$ pdffonts document.pdf
Syntax Error (5357): Bad block header in flate stream
Syntax Error (6710): Bad block header in flate stream
name                                 type              encoding         emb sub uni object ID
------------------------------------ ----------------- ---------------- --- --- --- ---------
Arial                                CID TrueType      Identity-H       no  no  yes      6  0
Arial,Bold                           CID TrueType      Identity-H       no  no  yes     10  0
```

Running gs from paperless container returns the same error above, but if I run it on my local opensuse box the file is rendered correctly with the following output

```
$ gs document.pdf
GPL Ghostscript 9.54.0 (2021-03-30)
Copyright (C) 2021 Artifex Software, Inc.  All rights reserved.
This software is supplied under the GNU AGPLv3 and comes with NO WARRANTY:
see the file COPYING for details.
Processing pages 1 through 1.
Page 1
Can't find CID font "Arial".
Attempting to substitute CID font /Adobe-Identity for /Arial, see doc/Use.htm#CIDFontSubstitution.
The substitute CID font "Adobe-Identity" is not provided either. attempting to use fallback CIDFont.See doc/Use.htm#CIDFontSubstitution.
Loading a TT font from /usr/share/ghostscript/9.54.0/Resource/CIDFSubst/DroidSansFallback.ttf to emulate a CID font Adobe-Identity ... Done.
Can't find CID font "Arial".
Attempting to substitute CID font /Adobe-Identity for /Arial, see doc/Use.htm#CIDFontSubstitution.
>>showpage, press <return> to continue<<
```

As per https://www.ghostscript.com/doc/9.53.0/Use.htm#CIDFontSubstitution when GS doesn't know what to do with a font it will fallback to DroidSansFallback.ttf.

If I run `apt install fonts-droid-fallback` inside the container then gs looks happier:

```
GPL Ghostscript 9.53.3 (2020-10-01)
Copyright (C) 2020 Artifex Software, Inc.  All rights reserved.
This software is supplied under the GNU AGPLv3 and comes with NO WARRANTY:
see the file COPYING for details.
Processing pages 1 through 1.
Page 1
Can't find CID font "Arial".
Attempting to substitute CID font /Adobe-Identity for /Arial, see doc/Use.htm#CIDFontSubstitution.
The substitute CID font "Adobe-Identity" is not provided either. attempting to use fallback CIDFont.See doc/Use.htm#CIDFontSubstitution.
Loading a TT font from /usr/share/ghostscript/9.53.3/Resource/CIDFSubst/DroidSansFallback.ttf to emulate a CID font Adobe-Identity ... Done.
Can't find CID font "Arial".
Attempting to substitute CID font /Adobe-Identity for /Arial, see doc/Use.htm#CIDFontSubstitution.
%%BoundingBox: 14 328 581 827
%%HiResBoundingBox: 14.490000 328.877990 580.809006 826.901975
```

but the output file is actually rendered with non-western characters, which of course confuses tesseract during the ocr phase.

The reason is that looking at `DroidSansFallback.ttf` in debian it only contains non-western characters while the same file inside the gs package from opensuse has the full character sets.

Providing the right `DroidSansFallback.ttf` would probably do the trick but I have no idea where to get it from debian repos.

**Comment by lemmi** [comment]
I also have a couple of problematic PDFs with the same result. 
I managed to open the broken PDF with `ghostscript` using [CID Font Substitution](https://www.ghostscript.com/doc/current/Use.htm#CIDFontSubstitution) and replace Arial with Liberation Sans. Having Microsoft Corefonts installed also works, but I was interested if there was a way without potential licensing issues:
```bash
# Make folder for an additional library search path
mkdir Init
# Add the manual CID Substitution (Format 2) to Init/cidfmap
echo '/Arial << /FileType /TrueType /Path (/usr/share/fonts/TTF/LiberationSans-Regular.ttf) /SubfontID 0 /CSI [(Identity) 0] >> ;' > Init/cidfmap
# Set GS_LIB environment to the extra configuration path
GS_LIB=./Init gs $THE_BROKEN_PDF
```

I noticed some issues with umlauts, but at least it mostly seems to work (it's fine with `Arial.TTF`)

This will also very likely need a suitable substitution for the other corefonts.

There is a debian package that ships a [downloader for Microsoft corefonts: ttf-mscorefonts-installer](https://packages.debian.org/stretch/ttf-mscorefonts-installer).

**Comment by lemmi** [comment]
Good news: I tried to embed the font and that also resolves the issue.
Assuming you have `Arial` installed locally under `/usr/share/fonts/TTF/Arial.TTF`, you can use something like this:
```bash
mkdir Init
echo '/Arial << /FileType /TrueType /Path (/usr/share/fonts/TTF/Arial.TTF) /CSI [(Identity) 0] >> ;' > Init/cidfmap
GS_LIB=./Init gs -dBATCH -sDEVICE=pdfwrite -o output.pdf input.pdf
```
Uploading the result seems to work without issues. So at least there is a workaround if there isn't a practical solution available to fix this in paperless.

**Comment by jhf2442** [comment]
@lemmi thanks for tracking this down

I think the most flexible thing would be to mount a directory with additional Fonts and cidfmap file into a standard location of paperless. This seems to be more flexible than patching the arial font into the software upon each release, and allows to add more fonts as needed

**Comment by lemmi** [comment]
Oh, I think that's a great idea. Since `ghostscript` can be instructed to use the `cidfmap` via environment variables, it should be an easy fix even now. This should be documented alongside a prepared `cidfmap` unless someone comes up with a better approach. If I don't forget, I should be able to prepare a `cidfmap` with all MS corefonts.

**Comment by jhf2442** [comment]
@lemmi Reminder :-)

**Comment by schnuffle** [comment]
Has anybody tried to use free fonts to see if they help?

There's the package "fonts-liberation" that offers free versions.
From the package description:

- `Description: Fonts with the same metrics as Times, Arial and Courier
 A set of serif, sans-serif and monospaced fonts from Red Hat with
 exactly the same metrics as the (non-free) Microsoft Times, Arial
 and Courier fonts, which implies those fonts can serve as a drop-in
 replacement. The font family is named Liberation.
`

**Comment by stumpylog** [comment]
It's now possible to run custom scripts at container startup.  This could help via installing more fonts or configuring Ghostscript in some way.

But if there's documentation updates as well, that would be great to have.

**Comment by jhf2442** [comment]
@stumpylog couldn't find it in the docker files, do you have a hint for me ? :-)

**Comment by stumpylog** [comment]
The docs won't be updated until the beta is pushed to main.  But you can see the general idea in #1838

**Comment by jhf2442** [comment]
ah OK, it's in the dev branch !  now I found it. reads good and looks like a very flexible way of applying patches in the future. Many thanks for this !

**Comment by shamoon** [comment]
Similarly to #1485 I think the startup scripts may be as close as we get with this, as noted we have to balance image size. Thanks all

**Comment by frrad** [comment]
Does anyone (@lemmi maybe?) have an example of a startup script which modifies the container's `cidfmap`? I have a bad pdf which I can render in `gs` outside of paperless by adding a custom font substitution rule to my ` `/var/lib/ghostscript/fonts/cidfmap`. It would be great to be able to do this in my paperless container though.

I'm likely to get more PDFs like this in the future and it would be nice to have it just work.

**Comment by lemmi** [comment]
I haven't had a look on how to run scripts, but to support ms fonts specifically I'd probably try and install the `ttf-mscorefonts-installer` package.

**Comment by frrad** [comment]
I did my own research, putting it down here for posterity:

Like the others on this thread I was having problems with a pdf missing some fonts that should have been embedded. In my case they were `ArialMT`, `Arial-BoldItalicMT`, `Arial-BoldMT` but in general you can read off the names by running `gs my_crap_pdf.pdf` and reading the errors as outlined in some of the previous comments.

Since my fonts are arial, and the docker container already installs the liberation fonts https://github.com/paperless-ngx/paperless-ngx/blob/7048af276ad25fc27a24a11369f4fc0fcc8ff2e4/Dockerfile#L92 I just had to figure out how to instruct gs to use them in the container. This is pretty easy with the custom scripts functionality described in `docs/advanced_usage.md`. Basically:

Register a new scripts volume in the container
````
    volumes:
...
      - ./scripts:/custom-cont-init.d:ro
```
making sure to set the permissions correctly as described in the docs.

Write a script say `fonts.sh` in your new `./scripts`
```
cat << EOF >> /var/lib/ghostscript/fonts/cidfmap
/ArialMT << /FileType /TrueType /Path (/usr/share/fonts/truetype/liberation/LiberationSans-Regular.ttf) /SubfontID 0 /CSI [(Identity) 0] >> ;
/Arial-BoldItalicMT << /FileType /TrueType /Path (/usr/share/fonts/truetype/liberation/LiberationSans-BoldItalic.ttf) /SubfontID 0 /CSI [(Identity) 0] >> ;
/Arial-BoldMT << /FileType /TrueType /Path (/usr/share/fonts/truetype/liberation/LiberationSans-Bold.ttf) /SubfontID 0 /CSI [(Identity) 0] >> ;
EOF
```

**Comment by mandre00** [comment]
@frrad Thanks for your post 
it worked for me too, I needed the font MSMincho which I mount into the container and then I needed to added an entry to `/var/lib/ghostscript/fonts/cidfmap`

```
cat << EOF >> /var/lib/ghostscript/fonts/cidfmap
/MSMincho << /FileType /TrueType /Path (/usr/share/fonts/CIDFont/MSMINCHO.TTF) /SubfontID 0 /CSI [(Identity) 0] >> ;
EOF
```

**Comment by jhf2442** [comment]
Same here, worked for me after I added definitions for Arial (w/o any suffix !) and Tahoma. No need to rework the dockerfile, therefore using the regular container from the repository. I've added a /fonts mountpoint to my docker-compose file just in case, in order to provide additional fonts if ever needed one day.
Looking at the PDF properties in okular helps to find the fonts being referenced
Many thanks @frrad for the template !
